### PR TITLE
People Action List prototype

### PR DESCRIPTION
**Pull request by howard-stearns**
Creates a People Action List with the same functionality as the mod.js script, per the prototype milestone of https://highfidelity.atlassian.net/wiki/display/HOME/People+Action+List

The core idea of the "PAL" is that you can click on avatar nodes to select in a list, or click in the list to show the selected avatar nodes. The list shows identifying information and options, that are stable as people move around.

The idea of the prototype is to have enough there there, such that we can have a stable base by which multiple can work on pieces that take this further, before it eventually replaces mod.js among defaultScripts.js.  The planned further work includes:

- not so hideously ugly as now.
- show username/machine-identification for admins.
- show status (including the showing of ignored people present)
- undo (i.e., toggles)
- show mic levels
- show entities lastEditedBy the selected user

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: b0a69468d22cfc686c448db4b9a426517eee39df available at: https://gist.github.com/4272af5ad3427bae47204ac7b22f34d7<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9202-b0a6946.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9202-b0a6946.dmg

**Comment by howard-stearns** [comment]
To operate: run scripts/system/pal.js to get a "People" button. Then press the button. Should be able to select (list row or avatar node overlay) with mouse or hand controllers. 
Note: list does not update as people come and go. Close the list and re-open.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 03d3e0b550fc8a2ba68f9219738764eb02ce8eea available at: https://gist.github.com/4272af5ad3427bae47204ac7b22f34d7<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9202-03d3e0b.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9202-03d3e0b.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b0547750b08b472f531850127b092bce7c38b7b3 available at: https://gist.github.com/4272af5ad3427bae47204ac7b22f34d7<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9202-b054775.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9202-b054775.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 773bbc4ef16e60559a37390193cb59a3eed613d8 available at: https://gist.github.com/4272af5ad3427bae47204ac7b22f34d7<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9202-773bbc4.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9202-773bbc4.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 00653b6b1910895bb42745e12c2e04a8a0e21284 available at: https://gist.github.com/4272af5ad3427bae47204ac7b22f34d7<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9202-00653b6.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9202-00653b6.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: a50b731faa2df90c8467d1cc0be142771a26827f available at: https://gist.github.com/4272af5ad3427bae47204ac7b22f34d7<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9202-a50b731.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9202-a50b731.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4503855ae8b02260f1aa32600e0683bc68d86de4 available at: https://gist.github.com/4272af5ad3427bae47204ac7b22f34d7<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9202-4503855.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9202-4503855.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d5c219a7616e61b50d8bb4c4590a2665a6bb98c0 available at: https://gist.github.com/4272af5ad3427bae47204ac7b22f34d7<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9202-d5c219a.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9202-d5c219a.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 66e2c651c63cc2ed61588b51f155ffa5fe3898ef available at: https://gist.github.com/4272af5ad3427bae47204ac7b22f34d7<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9202-66e2c65.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9202-66e2c65.dmg

File: scripts/system/pal.js [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+var triggerMapping = Controller.newMapping(Script.resolvePath('') + '-click');
+function controllerComputePickRay(hand) {
+    var controllerPose = getControllerWorldLocation(hand, true);
+    if (controllerPose.valid) {
+        return { origin: controllerPose.position, direction: Quat.getUp(controllerPose.orientation) };
+    }
+}
+function makeClickHandler(hand) {
+    return function (clicked) {
+        if (clicked > 0.85) {
```
Comment (zfox23) [inline scripts/system/pal.js]: Any reason this isn't a var like "UNSELECTED_COLOR" at the top?

Comment (howard-stearns) [inline scripts/system/pal.js]: Although most of our (older) scripts put constants at the top, our coding standard does specifically say to define them as near as practical to where they are used. As it happens, I agree with that particular part of the standard.

**Comment by howard-stearns** [comment]
test plan:
See "to operate" note, above. Go to a place with multiple people and ac-agents (protocol changes outside this pr require use of dev-xxx domains).
- All avatars (in including agents) should have a green ball, drawn in front of everything, and slightly transparent.
- Clicking the ball or the row in the list should select in both places. (highlighted in list, red ball in scene). Clicking either place should toggle off.
- Balls should go away and come back if you have the Personal-space Bubble on and go into the avatar, and then back off. (Note: list ONLY updates when you close and re-open.) When the ball comes back, it should be of the correct green/red color to reflect selected state.
- Closing the list (with the People toolbar button or with the close button on the People Action List frame) should be reflected in the toolbar button state.  Re-opening should get new list, but the list window should be in the same place as when it was closed (e.g. if it was moved).
-  Display names should be shown in the list, defaulted and uniquified in a crude way that will not necessarily be the same for everyone who opens the list.
- You should have an ignore button for each users. IFF you have "kick" rights in the domain, you should additionally have mute and ban buttons. All should work exactly as the mod.js script does. (The "ignore" or "kick" button in the toolbar.)

Know issues which will be worked on later, after merging. (Note that this prototype is not included in defaultScripts.js yet.):
- appearance
- some way to make clickable avatar nodes be visible in dark rooms, and prettier.
- display names should show the same sanitized, defaulted, uniqifiied names for each observer
- show usernames instead of the current fake session-id-based info, and only show if you have kick rights.
- when you ignore, it should continue to show in the list, and continue to show the ball (or replacement) until you close the list. You should be able to undo the ignore by re-clicking.

**Comment by zfox23** [comment]
*QA: SOME FAILURES - see notes below* Win10/Rift/Hydra, tested on desktop and HMD.

> All avatars (in including agents) should have a green ball, drawn in front of everything, and slightly transparent.
Pass. Understood that these can't be fullbright.

> Clicking the ball or the row in the list should select in both places. (highlighted in list, red ball in scene). Clicking either place should toggle off.
Clicking the ball does indeed toggle on/off. *However,* clicking a name in the list doesn't toggle; clicking a name selects but never deselects. I'm unsure if you intended this (and just mistyped in the test plan)

> Balls should go away and come back if you have the Personal-space Bubble on and go into the avatar, and then back off. (Note: list ONLY updates when you close and re-open.) When the ball comes back, it should be of the correct green/red color to reflect selected state.
Fail. If I select an avatar, then get close to it with my space bubble enabled, then back off, it does indeed remain selected; this part passes. However, if I then select another avatar (either by clicking on the ball or on its name in the list), the original avatar's bubble remains red.

> Closing the list (with the People toolbar button or with the close button on the People Action List frame) should be reflected in the toolbar button state. Re-opening should get new list, but the list window should be in the same place as when it was closed (e.g. if it was moved).
Pass.

> Display names should be shown in the list, defaulted and uniquified in a crude way that will not necessarily be the same for everyone who opens the list.
Pass.

> You should have an ignore button for each users. IFF you have "kick" rights in the domain, you should additionally have mute and ban buttons. All should work exactly as the mod.js script does. (The "ignore" or "kick" button in the toolbar.)
Pass. Limitation understood that banning an AC script that belongs to me will also ban me.

**Comment by howard-stearns** [comment]
And another bug: with that last change, I made the overlays disappear with you drive, and then not come back. Oops.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0e81e713255143731d2b429244d0a3b588a603ac available at: https://gist.github.com/4272af5ad3427bae47204ac7b22f34d7<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9202-0e81e71.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9202-0e81e71.dmg

**Pull request merged**